From 479e3f6efdba0517ec8f7ea98419e860ab44dc80 Mon Sep 17 00:00:00 2001
From: Gleb Mazovetskiy <glex.spb@gmail.com>
Date: Mon, 27 Apr 2020 00:06:18 +0100
Subject: [PATCH] Revert "The XBox One S controller sends keys outside the
 standard joystick button range"

This reverts commit c0bf85bee4bdc7cfeadccbf2e2a426d11620a357.

This commit introduced an issue with inputs registering both as keyboard
keys and joystick buttons when using linkdev joystick mode.

See https://bugzilla.libsdl.org/show_bug.cgi?id=5058
---
 src/joystick/linux/SDL_sysjoystick.c   | 15 +++++++++------
 src/joystick/linux/SDL_sysjoystick_c.h |  2 +-
 2 files changed, 10 insertions(+), 7 deletions(-)

diff --git a/src/joystick/linux/SDL_sysjoystick.c b/src/joystick/linux/SDL_sysjoystick.c
index 28bccd8ea..f17baebb2 100644
--- a/src/joystick/linux/SDL_sysjoystick.c
+++ b/src/joystick/linux/SDL_sysjoystick.c
@@ -655,16 +655,16 @@ ConfigJoystick(SDL_Joystick * joystick, int fd)
 #ifdef DEBUG_INPUT_EVENTS
                 printf("Joystick has button: 0x%x\n", i);
 #endif
-                joystick->hwdata->key_map[i] = joystick->nbuttons;
+                joystick->hwdata->key_map[i - BTN_MISC] = joystick->nbuttons;
                 ++joystick->nbuttons;
             }
         }
-        for (i = 0; i < BTN_JOYSTICK; ++i) {
+        for (i = BTN_MISC; i < BTN_JOYSTICK; ++i) {
             if (test_bit(i, keybit)) {
 #ifdef DEBUG_INPUT_EVENTS
                 printf("Joystick has button: 0x%x\n", i);
 #endif
-                joystick->hwdata->key_map[i] = joystick->nbuttons;
+                joystick->hwdata->key_map[i - BTN_MISC] = joystick->nbuttons;
                 ++joystick->nbuttons;
             }
         }
@@ -964,9 +964,12 @@ HandleInputEvents(SDL_Joystick * joystick)
             code = events[i].code;
             switch (events[i].type) {
             case EV_KEY:
-                SDL_PrivateJoystickButton(joystick,
-                                          joystick->hwdata->key_map[code],
-                                          events[i].value);
+                if (code >= BTN_MISC) {
+                    code -= BTN_MISC;
+                    SDL_PrivateJoystickButton(joystick,
+                                              joystick->hwdata->key_map[code],
+                                              events[i].value);
+                }
                 break;
             case EV_ABS:
                 switch (code) {
diff --git a/src/joystick/linux/SDL_sysjoystick_c.h b/src/joystick/linux/SDL_sysjoystick_c.h
index a92429bcc..1903bdf32 100644
--- a/src/joystick/linux/SDL_sysjoystick_c.h
+++ b/src/joystick/linux/SDL_sysjoystick_c.h
@@ -51,7 +51,7 @@ struct joystick_hwdata
     } *balls;
 
     /* Support for the Linux 2.4 unified input interface */
-    Uint8 key_map[KEY_MAX];
+    Uint8 key_map[KEY_MAX - BTN_MISC];
     Uint8 abs_map[ABS_MAX];
     struct axis_correct
     {
-- 
2.25.1

